### PR TITLE
rad phase 2: hc-happ-scaffold

### PR DESCRIPTION
**Pull request by JettTech**
## 1 - Get Holonix and enter the development environment
Holonix is a full Holochain development environment built with the [Nix package manager](https://nixos.org/nix/). 
This repo comes shipped with nix-shell already baked in.  All you need to do, is enter the provided environment, by running:
```
$ nix-shell
```
## 2 - Run the hc-happ-scaffold command
From within the `nix-shell` environment, first create a directory for all your Holochain projects (if you haven't already). You can create it wherever you like; here's a recommended setup:

```
$ cd ~
$ mkdir Holochain
$ cd Holochain
```

Then run this command:

```
$ hc-happ-scaffold <PATH_TO_JSON_TYPE_SPEC> <PATH_TO_NEW_HAPP_DIRECTORY>
```

This will create a new directory for your project, download all the dependencies and development tools, and create the hApp source code.

## 3 - Start your new happ!

Once it's complete, go into the new project directory and run this command:

```
$ npm run start
```

### CODE REVIEW

**Comment by robbiecarlton** [comment]
@JettTech this will need a rewrite with the changes we made to the structure of the rad tools

**Comment by JettTech** [comment]
@robbiecarlton @zo-el - this is ready for your review again, as all the structural updates to RAD Tools Phase 2 is complete - at least for take 1. :+1:

File: happs/scaffold/default.nix [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  ''
+    ''${1?"Command Usage Error: ARG 1 - PATH TO SCHEMA REQUIRED"}
+    [[ $(mimetype -b "''${1}") != "application/json" ]] && { echo "Command Usage Error: ARG 1 - JSON FILE TYPE REQUIRED"; exit 1; }
+    STARTING_DIRECTORY=$(pwd -P)
+    curl -L -o happ-scaffold.tar.gz https://github.com/holochain/RAD-Tools-Phase-2/archive/integration-testing-tryorama-fix-1.tar.gz
+    mkdir "''${2:-"My-New-App"}"
+    tar -zxvf happ-scaffold.tar.gz --strip-components=1 -C ./"''${2:-"My-New-App"}"
+    rm happ-scaffold.tar.gz
+    cd ''${2:-"My-New-App"}
+    sed -i "s/RAD-Tools-Phase-2/"''${2:-"My-New-App"}"/g" package.json
```
Comment (robbiecarlton) [inline happs/scaffold/default.nix]: `RAD-Tools-Phase-2` should be all lowercase now

**Review by robbiecarlton — DISMISSED** [review]

File: happs/scaffold/default.nix (hunk 1 above)
Comment (robbiecarlton) [reply happs/scaffold/default.nix]: actually, let's have this string be something more clearly a template, like `<HAPP-NAME>`

Comment (JettTech) [reply happs/scaffold/default.nix:17]: Done.  - FYI: This is a change in the RAD tool repo itself, so the change won't be reflected here.

**Comment by thedavidmeister** [comment]
@JettTech can you point me at a json spec file i can use for testing?

**Comment by JettTech** [comment]
@thedavidmeister @joel - here is the link to a type-spec.json reference for testing:
https://gist.github.com/JettTech/a27fdad85f2b8f69182f5318d5b14e19

**Review by zo-el — APPROVED** [review]

File: happs/scaffold/default.nix [hunk 2]
```diff
@@ -0,0 +1,31 @@
+{ pkgs }:
+let
+  # ie: hc-happ-scaffold <path-to-package-file> <path-to-new-happ-dir>
+  name = "hc-happ-scaffold";
+
+  script = pkgs.writeShellScriptBin name
+  ''
+    ''${1?"Command Usage Error: ARG 1 - PATH TO SCHEMA REQUIRED"}
+    [[ $(mimetype -b "''${1}") != "application/json" ]] && { echo "Command Usage Error: ARG 1 - JSON FILE TYPE REQUIRED"; exit 1; }
```
Comment (thedavidmeister) [inline happs/scaffold/default.nix]: can't assume `mimetype` exists, can we add it to the build inputs?

File: happs/scaffold/default.nix [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    mkdir "''${2:-"My-New-App"}"
+    tar -zxvf happ-scaffold.tar.gz --strip-components=1 -C ./"''${2:-"My-New-App"}"
+    rm happ-scaffold.tar.gz
+    cd ''${2:-"My-New-App"}
+    [ ! -d "./src/setup" ] && mkdir ./src/setup
+    [ ! -d "./src/setup/type-specs" ] && mkdir ./src/setup/type-specs
+    [ -f "./src/setup/type-specs/type-spec.json" ] && rm -rf ./src/setup/type-specs/type-spec.json
+    curl -L -o ./src/setup/type-specs/type-spec.json "file://$(readlink -f $(realpath --relative-to=$(pwd -P) $STARTING_DIRECTORY)/''${1})"
+    [ ! -f "./src/setup/type-specs/sample-type-spec.json" ] &&  cp ./src/setup/type-specs/type-spec.json ./src/setup/type-specs/sample-type-spec.json
+    npm i
```
Comment (thedavidmeister) [inline happs/scaffold/default.nix:33]: should we be using yarn?

File: happs/scaffold/default.nix [hunk 4]
```diff
@@ -0,0 +1,31 @@
+{ pkgs }:
+let
+  # ie: hc-happ-scaffold <path-to-package-file> <path-to-new-happ-dir>
+  name = "hc-happ-scaffold";
+
+  script = pkgs.writeShellScriptBin name
+  ''
+    ''${1?"Command Usage Error: ARG 1 - PATH TO SCHEMA REQUIRED"}
```
Comment (thedavidmeister) [inline happs/scaffold/default.nix:8]: @JettTech i was hoping we could point at URLs as well as paths for the whole process, here's a one liner i tried:

```
nix-shell https://github.com/holochain/holonix/archive/2020-04-09/rad-phase2.tar.gz --run 'hc-happ-scaffold https://tinyurl.com/ybgdhtwa $TMP/bar && cd $TMP/bar && npm run start'
```

the tiny url points to the gist you linked above

doesn't work it seems to need a local path

**Review by thedavidmeister — CHANGES_REQUESTED** [review]

**Comment by thedavidmeister** [comment]
@JettTech also can we get some tests written for this?

File: happs/scaffold/default.nix (hunk 3 above)
Comment (JettTech) [reply happs/scaffold/default.nix:33]: As a team, we had decided to use npm.  Do you have a strong opinion in favor for yarn?  Open to hearing your thoughts.

File: happs/scaffold/default.nix (hunk 4 above)
Comment (JettTech) [reply happs/scaffold/default.nix:8]: This is correct, the expected behavior is for it to build based on a reference to a local file path. Generally, the goal was to keep things as simple as possible for the first version, but I do agree that it'd be best if the path could be a url or file reference.  

I'll look into this. :+1:

**Review by thedavidmeister — APPROVED** [review]
i approve once the tests pass :)

**Review by zo-el — APPROVED** [review]

**Pull request merged**